### DISCUSSION TITLE
DB connection large delay with 2.9.x

### DISCUSSION THREAD

**Discussion by december1981** [psycopg/psycopg2#1423] [General]
When I disconnect from my VPN software, reconnecting to the database with a 2.9.x version of psycopg2 hangs for some time, before duly connecting (certainly frequently, if not 100%). This does not happen with 2.8.6.

This is not necessarily a problem with psycopg2 - but to help me get to the bottom of this, can anyone think of what might have changed between these major versions which could cause this?

**Comment by dvarrazzo** [comment]
I guess you use Windows?

**Reply by december1981** [reply]
Sorry, should have mentioned - this is with Ubuntu 20.

**Reply by dvarrazzo** [reply]
You can try to check it using psycopg 3. Add a print [here around](https://github.com/psycopg/psycopg/blob/f531327eb50324cb87aca068fb404adb905f14d3/psycopg/psycopg/generators.py#L39) to check the status during connection time: it might suggest where it gets stuck.

**Reply by december1981** [reply]
It seems to delay when conn.status is

	`CONNECTION_GSS_STARTUP,		/* Negotiating GSSAPI. */`

(Looking at libpq-fe.h ConnStatusType enum)

and when conn.connect_poll() is then called, the long delay occurs before the python code moves on.

**Reply by december1981** [reply]
conn.status had a value of 11, anyway.

**Reply by dvarrazzo** [reply]
Yes, that's the thing I've seen it happening often, but always on Windows. `gssencmode=disable` works around it.

I don't know why that happens: it seems to depend on the libpq. Never get any meaningful reply whenever I asked.

**Reply by december1981** [reply]
Yep, that fixed it. Good to know.

**Reply by dvarrazzo** [reply]
Writing it as a reply to accept, so we play little stackoverflow...

**Comment by dvarrazzo** [comment]
This seems to happen often with the GSSAPI negotiation.

If you don't need GSSAPI, try using `gssencmode=disable` in your connection string.
